### PR TITLE
Added a switcher for turning off write controls for read-only users a…

### PR DESCRIPTION
**Pull request by hsiang9431-amzn**
…nd tenants

Signed-off-by: David Bennett <david.bennett@eliatra.com>

##  opensearch-project/security-dashboards-plugin pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Bug Fix
 
 
 2. __Github Issue # or road-map entry, if available:__
https://github.com/opensearch-project/security-dashboards-plugin/issues/776

 3. __Description of changes:__

A user might not have sufficient permissions to make any changes to saved objects like dashboards, visualization etc. This can happen for two reasons: a) the user has a Kibana read-only role and thus can only see the Dashboards, but is not able to modify anything. b) the currently selected tenant is read-only for the user. In these cases, displaying write controls like "save" or "edit" buttons do not make sense. This PR introduces a switcher which will turn off all known write controls for these cases:

In Dashboards you can register a so-called switcher on the Capabilities. Capabilities define what features (or "capabilities") of a plugin are enabled or disabled. This PR implements and registers a switcher. It will first evaluate whether the user us currently in read-only mode or not. If so, the switcher will deactivate all known write controls.

 4. __Why these changes are required?__ 
If a user is in read-only mode, it does not make sense to offer any write controls.

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)
Before: Write controls would be displayed. After: Write controls are disabled.

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 Tested the changes manually, added jest unit tests in read_only_capabilities_switcher.test.ts.
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

This PR only covers "known" capabilities from the standard installation. However, any plugin (including third-party plugins) can make use of the capabilities feature. Since we do not know what these capabilities from third-parties look like in advance, and since there is also no way of defining which capability is a "write control", there is currently no way of automatically disable write controls for third party plugins. We would need a system where a plugin could explicitly  specify its write controls.  However, this is beyond the scope of this PR and might need a broader discussion.

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)
No.

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/plugin.ts [hunk 1]
```diff
@@ -43,6 +43,7 @@ import { getAuthenticationHandler } from './auth/auth_handler_factory';
 import { setupMultitenantRoutes } from './multitenancy/routes';
 import { defineAuthTypeRoutes } from './routes/auth_type_routes';
 import { createMigrationOpenSearchClient } from '../../../src/core/server/saved_objects/migrations/core';
+import { ReadOnlyCapabilitiesSwitcher } from './auth/read_only_capabilities_switcher';
```
Comment (davidlago) [inline server/plugin.ts:46]: Why do we have a different capitalization convention for this one? it seems like every other one imported above starts with lower case

File: server/auth/read_only_capabilities_switcher.test.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  catalogue: {},
+};
+
+describe('test read-only role evaluation', () => {
+  const authinfo = {
+    roles: ['roleA', 'roleB'],
+  };
+
+  const authinfoDefaultReadonlyRoles = {
+    roles: ['roleA', 'kibana_read_only'],
```
Comment (davidlago) [inline server/auth/read_only_capabilities_switcher.test.ts:42]: If we extract that list of default roles to a more global context as suggested by my other comment, please modify this test file to reference those as opposed to hardcoding them here

**Review by davidlago — CHANGES_REQUESTED** [review]
Have we tested this manually?

**Comment by hsiang9431-amzn** [comment]
> Have we tested this manually?

I built from the branch and spun up a local cluster, then check its behavior

**Review by davidlago — DISMISSED** [review]

**Review by nsri19 — APPROVED** [review]

**Review by davidlago — DISMISSED** [review]

**Comment by hsiang9431-amzn** [comment]
Please review Dave's comments on this PR https://github.com/opensearch-project/security-dashboards-plugin/pull/867

**Comment by cliu123** [comment]
I've rebased the PR against `main` branch. 
@opensearch-project/security Please review.

**Comment by cliu123** [comment]
Closing the PR due to inactivity for now. Please feel free to re-open if needed.

**Pull request closed without merging**